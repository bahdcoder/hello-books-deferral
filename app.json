{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
