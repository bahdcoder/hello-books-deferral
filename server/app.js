import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
