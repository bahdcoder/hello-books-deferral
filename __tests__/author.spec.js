import supertest from 'supertest';
import { app, databaseConnection } from '@server/app';

const server = () => supertest(app);
let author;

describe('AUTHOR API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection.migrate.latest();
        await databaseConnection('authors').truncate();
    });

    beforeEach(() => {
        author = { name: 'john doe' };
    });

    afterAll(async () => {
        await databaseConnection('authors').truncate();
    });

    describe('POST CREATE AUTHOR api/v1/authors', () => {
        it('should not create author if name field is empty', async () => {
            const { status, body } = await server()
                .post('/api/v1/authors')
                .send({
                    ...author,
                    name: ''
                });

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should create author with valid inputs', async () => {
            const { status, body } = await server()
                .post('/api/v1/authors')
                .send(author);

            expect(status).toBe(201);
            expect(body).toMatchSnapshot();
        });
    });

    describe('GET AUTHOR api/v1/authors', () => {
        it('should not get author if id param is not a number', async () => {
            const { status, body } = await server().get('/api/v1/authors/d');

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should return single author with name query string', async () => {
            const { status, body } = await server().get('/api/v1/authors/1');

            expect(status).toBe(200);
            expect(body.data.name).toBe(author.name);
            expect(body).toMatchSnapshot();
        });
    });

    describe('UPDATE AUTHOR api/v1/authors', () => {
        it('should not update author if name field is empty', async () => {
            const { status, body } = await server()
                .patch('/api/v1/authors/1')
                .send({
                    ...author,
                    name: ''
                });

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should update author with valid inputs', async () => {
            const { status, body } = await server()
                .patch('/api/v1/authors/1')
                .send({
                    ...author,
                    name: 'Jane Doe'
                });

            expect(status).toBe(200);
            expect(body.data.name).toBe('Jane Doe');
            expect(body).toMatchSnapshot();
        });
    });

    describe('DELETE AUTHOR api/v1/authors', () => {
        it('should not delete author if id param is not a number', async () => {
            const { status, body } = await server().delete('/api/v1/authors/d');

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should delete author with id param valid', async () => {
            const { status, body } = await server().delete('/api/v1/authors/1');

            expect(status).toBe(200);
            expect(body).toMatchSnapshot();
        });
    });
});
