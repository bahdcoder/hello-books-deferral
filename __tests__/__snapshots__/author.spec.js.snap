// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AUTHOR API ENDPOINTS DELETE AUTHOR api/v1/authors should delete author with id param valid 1`] = `
Object {
  "data": Object {
    "message": "Author deleted successfully",
  },
  "status": "success",
}
`;

exports[`AUTHOR API ENDPOINTS DELETE AUTHOR api/v1/authors should not delete author if id param is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "id",
      "message": "id must be an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS GET ALL AUTHOR api/v1/authors should get all authors 1`] = `
Object {
  "data": Object {
    "results": Array [
      Object {
        "id": 3,
        "name": "James Bond",
      },
    ],
    "total": 7,
  },
  "status": "success",
}
`;

exports[`AUTHOR API ENDPOINTS GET ALL AUTHOR api/v1/authors should not get author if limit query is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "page",
      "message": "page is expected to be a an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS GET ALL AUTHOR api/v1/authors should not get author if limit query is not a number 2`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "limit",
      "message": "limit is expected to be a an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS GET ALL AUTHOR api/v1/authors should not get author if page query is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "page",
      "message": "page is expected to be a an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS GET AUTHOR api/v1/authors should not get author if id param is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "id",
      "message": "id must be an integer",
      "validation": "number",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS GET AUTHOR api/v1/authors should return single author with name query string 1`] = `
Object {
  "data": Object {
    "id": 1,
    "name": "John Doe",
  },
  "status": "success",
}
`;

exports[`AUTHOR API ENDPOINTS POST CREATE AUTHOR api/v1/authors should create author with valid inputs 1`] = `
Object {
  "data": Object {
    "id": 5,
    "name": "John Doe",
  },
  "status": "success",
}
`;

exports[`AUTHOR API ENDPOINTS POST CREATE AUTHOR api/v1/authors should not create author if name field is empty 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "name",
      "message": "name is required to create an author",
      "validation": "required",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS UPDATE AUTHOR api/v1/authors should not update author if name field is empty 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "name",
      "message": "name is required to update an author",
      "validation": "required",
    },
  ],
  "status": "error",
}
`;

exports[`AUTHOR API ENDPOINTS UPDATE AUTHOR api/v1/authors should update author with valid inputs 1`] = `
Object {
  "data": Object {
    "id": 1,
    "name": "Jane Doe",
  },
  "status": "success",
}
`;
